src/Microsoft.Extensions.Primitives/CompositeChangeToken.cs

```diff
@@ -0,0 +1,138 @@
+{
+    /// <summary>
+    /// An <see cref="IChangeToken"/> which represents one or more <see cref="IChangeToken"/> instances.
+    /// </summary>
+    public class CompositeChangeToken : IChangeToken
+    {
+        private static readonly Action<object> _onChangeDelegate = OnChange;
+        public CancellationTokenSource _cancellationTokenSource;
+        private readonly object _callbackLock = new object();
+        private bool _registeredCallbackProxy;
+        private List<IDisposable> _disposables;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="CompositeChangeToken"/>.
+        /// </summary>
+        /// <param name="changeTokens">The list of <see cref="IChangeToken"/> to compose.</param>
+        public CompositeChangeToken(IReadOnlyList<IChangeToken> changeTokens)
+        {
+            if (changeTokens == null)
+            {
+                throw new ArgumentNullException(nameof(changeTokens));
+            }
+
+            ChangeTokens = changeTokens;
+
+            ActiveChangeCallbacks = false;
+            for (var i = 0; i < ChangeTokens.Count; i++)
+            {
+                if (ChangeTokens[i].ActiveChangeCallbacks)
+                {
+                    ActiveChangeCallbacks = true;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the list of <see cref="IChangeToken"/> which compose the current <see cref="CompositeChangeToken"/>.
+        /// </summary>
+        public IReadOnlyList<IChangeToken> ChangeTokens { get; }
+
+        /// <inheritdoc />
+        public IDisposable RegisterChangeCallback(Action<object> callback, object state)
+        {
+            EnsureCallbacksInitialized();
+            if (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                return _cancellationTokenSource.Token.Register(callback, state);
+            }
+
+            else
+            {
+                return null;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool HasChanged
+        {
+            get
+            {
+                if (_cancellationTokenSource.Token.IsCancellationRequested)
+                {
+                    return true;
+                }
+
+                for (var i = 0; i < ChangeTokens.Count; i++)
+                {
+                    if (ChangeTokens[i].HasChanged)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool ActiveChangeCallbacks { get; }
+
+        private void EnsureCallbacksInitialized()
+        {
+            if (_registeredCallbackProxy)
+            {
+                return;
+            }
+
+            lock (_callbackLock)
+            {
+                if (_registeredCallbackProxy)
+                {
+                    return;
+                }
+
+                _registeredCallbackProxy = true;
```

Comment:
Have to move this to end of the method if you're lazily creating `CancellationTokenSource`.